tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -375,72 +370,62 @@ Status TryRewriteForFinalOutput(ConvWithConvertOrClamp match) {
-    // skipped by TryRewriteForConvertOrClampImpl
-    match.convert_or_clamp = match.convert_or_clamp->users()[0];
-  } else if (match.conv->operand_count() >= 4 &&
-             is_one_to_one_X_to_Y_cast(convert, F32, S8) &&
-             is_one_to_one_X_to_Y_cast(match.conv->operand(3), S8, F32)) {
-    // If both conv output and side_input have convert_float_to_int8 and
-    // convert_int8_to_float, respectively, absorb both.
-    //
-    // Update convert_or_clamp with the output converter, so that it will be
-    // skipped by TryRewriteForConvertOrClampImpl
-    match.convert_or_clamp = match.convert_or_clamp->users()[0];
-    // Bypass the converter on side input
+    // Side input doesn't exist in this case.
+    TF_ASSIGN_OR_RETURN(changed, TryRewriteForConvertOrClampImpl(match));
+  } else if (is_one_to_one_X_to_Y_cast(match.conv->operand(3), S8, F32)) {
+    // If side_input has a convert_float_to_int8, absorb it as well.
     auto side_converter = match.conv->mutable_operand(3);
     side_converter->ReplaceAllUsesWithDifferentShape(
         side_converter->mutable_operand(0));
     side_converter->parent()->RemoveInstructionAndUnusedOperands(
         side_converter);
-  }
 
-  return TryRewriteForConvertOrClampImpl(match);
+    TF_ASSIGN_OR_RETURN(changed, TryRewriteForConvertOrClampImpl(match));
+  }
+  return changed;
 }
 
 // Fuse the clamp/convert pattern with the int8 convolution custom call
 // (either pure or fused) for int8 output
-StatusOr<bool> RunFuseClamp(
-    HloModule* module,
-    std::unordered_set<const HloInstruction*>& tracked_custom_calls) {
+StatusOr<bool> RunFuseClamp(HloModule* module) {
   bool changed = false;
   for (HloComputation* computation : module->MakeNonfusionComputations()) {
     std::vector<ConvWithConvertOrClamp> matches;
     int num_forward_convs = 0;
     for (auto instr : computation->instructions()) {
-      auto match = FindConvWithClamp(instr);
+      auto match = FindConvWithClampAndConvertToInt8(instr);
       if (match.has_value()) {
         matches.push_back(*match);
       }
-      if (auto call = DynCast<HloCustomCallInstruction>(instr)) {
-        if (call->custom_call_target() == kCudnnConvForwardCallTarget) {
-          num_forward_convs++;
-        }
-      }
     }
-    VLOG(1) << "Identified cuDNN forward conv + convert: " << matches.size()
-            << " out of " << num_forward_convs << " forward convs.";
     std::vector<std::pair<HloInstruction*, std::unique_ptr<HloInstruction>>>
         replacements;
     for (const ConvWithConvertOrClamp& match : matches) {
-      TF_RETURN_IF_ERROR(TryRewriteForFinalOutput(match));
-      // We have done with the convolution custom_call.
-      tracked_custom_calls.erase(match.conv);
-      changed = true;
+      TF_ASSIGN_OR_RETURN(bool c, TryRewriteForFinalOutput(match));
+      changed |= c;
+    }
+
+    // Report error for any convolution still having int32 output.
```

Comment:
I have tried not having this piece of code.  int32 output convolution today will be flagged by BuffereComparator::CompareEqual, because S32 is not implemented.  Once BuffereComparator::CompareEqual supports S32, cudnn_conv_runner may be the next one checking against int32 output convolution.  If no sanity check flags it, CuDNN itself will error about it eventually.  However, I prefer raise error in this function, because it is designed to handle int32 output convolution for CuDNN and is able to give specific message.